File: src/lua/clue/libc.lua
-- clue.crt
--
-- Clue C libc
--
-- © 2008 David Given.
-- Clue is licensed under the Revised BSD open source license. To get the
-- full license text, see the README file.
--
-- $Id$

require "clue.crt"
require "socket"

local unpack = unpack
local type = type
local print = print

local ptrread = clue.crt.ptrread
local ptrwrite = clue.crt.ptrwrite
local ptroffset = clue.crt.ptroffset
local ptrtostring = clue.crt.ptrtostring
local newptr = clue.crt.newptr
local math_floor = math.floor
local tonumber = tonumber
local io_write = io.write
local io_stdin = io.stdin
local io_stdout = io.stdout
local io_stderr = io.stderr
local string_format = string.format
local string_gsub = string.gsub
local math_sin = math.sin
local math_cos = math.cos
local math_sqrt = math.sqrt
local math_pow = math.pow
local math_log = math.log
local math_atan = math.atan
local math_exp = math.exp
local socket_gettime = socket.gettime

module "clue.libc"

function _malloc(sp, stack, size)
	return 1, {}
end

function _calloc(sp, stack, size1, size2)
	local d = {}
	for i = 1, (size1*size2) do
		d[i] = 0
	end
	return 1, d
end
	
function _free(sp, stack, po, pd)
end

function _realloc(sp, stack, po, pd)
	return po, pd
end

function _memcpy(sp, stack, destpo, destpd, srcpo, srcpd, count)
	for offset = 0, count-1 do
		destpd[destpo+offset] = srcpd[srcpo+offset]
	end
	return destpo, destpd
end

function _memset(sp, stack, destpo, destpd, c, n)
	for offset = 0, n-1 do
		destpd[destpo+offset] = c
	end
	return destpo, destpd
end
	
function _atoi(sp, stack, po, pd)
	local s = ptrtostring(po, pd)
	return math_floor(tonumber(s))
end

_atol = _atoi

function _printf(sp, stack, formatpo, formatpd, ...)
	format = ptrtostring(formatpo, formatpd)
	local inargs = {...}
	local outargs = {}
	
	local inargcount = #inargs
	local i = 1
	while (i <= inargcount) do
		local thisarg = inargs[i]
		i = i + 1
		if (i <= inargcount) then
			local nextarg = inargs[i]
			if (nextarg == nil) or (type(nextarg) == "table") then
				-- If the next argument is nil or a table, then we must
				-- be looking at a register pair representing a pointer.
				thisarg = ptrtostring(thisarg, nextarg)
				i = i + 1
			end
		end
		
		outargs[#outargs+1] = thisarg
	end	

	-- Massage the format string to be Lua-compatible.
	
	format = string_gsub(format, "[^%%]%%l", "%%")
	format = string_gsub(format, "^%%l", "%%")
	
	-- Use Lua's string.format to actually do the rendering.
	
	io_write(string_format(format, unpack(outargs)))
	return 1
end

__stdin = io_stdin
__stdout = io_stdout
__stderr = io_stderr
 
function _putc(sp, stack, c, fppo, fppd)
	fppd:write(string_format("%c", c))
end

-----------------------------------------------------------------------------
--                                STRINGS                                  --
-----------------------------------------------------------------------------

function _strcpy(sp, stack, destpo, destpd, srcpo, srcpd)
	local origdestpo = destpo
	
	while true do
		local c = srcpd[srcpo]
		destpd[destpo] = c
		
		srcpo = srcpo + 1
		destpo = destpo + 1
		
		if (c == 0) then
			break
		end
	end
	
	return origdestpo, destpd
end

-----------------------------------------------------------------------------
--                                 MATHS                                   --
-----------------------------------------------------------------------------

function _sin(sp, stack, x)
	return math_sin(x)
end

function _cos(sp, stack, x)
	return math_cos(x)
end

function _sqrt(sp, stack, x)
	return math_sqrt(x)
end

function _pow(sp, stack, x, y)
	return math_pow(x, y)
end

function _log(sp, stack, x)
	return math_log(x)
end

function _atan(sp, stack, x)
	return math_atan(x)
end

function _exp(sp, stack, x)
	return math_exp(x)
end

-----------------------------------------------------------------------------
--                                 TIME                                    --
-----------------------------------------------------------------------------

function _gettimeofday(sp, stack, tvpo, tvpd, tzpo, tzpd)
	local t = socket_gettime()
	local secs = math_floor(t)
	local usecs = math_floor((t - secs) * 1000000)
	tvpd[tvpo+0] = secs
	tvpd[tvpo+1] = usecs
	return 0
end

File: src/perl5/libc.pl
# Perl5 C runtime
#
# © 2008 David Given.
# Clue is licensed under the Revised BSD open source license. To get the
# full license text, see the README file.
#
# $Id: cluerun 20 2008-07-16 10:21:58Z dtrg $
# $HeadURL: https://cluecc.svn.sourceforge.net/svnroot/cluecc/clue/cluerun $
# $LastChangedDate: 2008-07-16 11:21:58 +0100 (Wed, 16 Jul 2008) $

use Time::HiRes qw(gettimeofday);

#############################################################################
#                                 MEMORY                                    #
#############################################################################

$_malloc = sub
{
	return 0, [];
};

$_calloc = sub
{
	my ($stackpo, $stackpd, $s1, $s2) = @_;
	my @d = ();
	
	for $i (0 .. ($s1 * $s2))
	{
		$d[$i] = 0;
	}
	
	return 0, \@d;
};

#############################################################################
#                                 STRINGS                                   #
#############################################################################

$_strcpy = sub
{
	my ($stackpo, $stackpd, $destpo, $destpd, $srcpo, $srcpd) = @_;
	my $origdestpo = $destpo;
	
	my $c;
	do
	{
		$c = $srcpd->[$srcpo];
		$destpd->[$destpo] = $c;
		$srcpo++;
		$destpo++;
	}
	while ($c != 0);
	
	return $origdestpo, $destpd;
};
 	
$_memset = sub
{
	my ($stackpo, $stackpd, $destpo, $destpd, $c, $n) = @_;
	my $origdestpo = $destpo;
	
	while ($n > 0)
	{
		$destpd->[$destpo] = $c;
		$n--;
	}
	
	return $origdestpo, $destpd;
};

#############################################################################
#                                  STDIO                                    #
#############################################################################

$| = 1; # force flush on every write

$_printf = sub
{
	my ($stackpo, $stackpd, $formatpo, $formatpd) = @_;
	my $format = clue_ptr_to_string($formatpo, $formatpd);
	my @outargs = ();
	
	my $numargs = $#_;
	my $i = 4;
	while ($i <= $numargs)
	{
		my $thisarg = $_[$i];
		$i++;
		if ($i < $numargs)
		{
			my $nextarg = $_[$i];
			if (ref $nextarg)
			{
				# If the next argument is NULL or a table, then we must
				# be looking at a register pair representing a pointer.
				$thisarg = clue_ptr_to_string($thisarg, $nextarg);
				$i++;
			}
		}
		
		push @outargs, $thisarg;
	}
	
	print sprintf($format, @outargs);
	
	return 1;
};

$_putc = sub
{
	my ($stackpo, $stackpd, $c, $fppo, $fppd) = @_;
	printf sprintf("%c", $c);
};

$_atoi = sub
{
	my ($stackpo, $stackpd, $srcpo, $srcpd) = @_;
	my $s = clue_ptr_to_string($srcpo, $srcpd);
	return $s | 0;
};

$_atol = $_atoi;
	
#############################################################################
#                                   TIME                                    #
#############################################################################

$_gettimeofday = sub
{
	my ($sp, $stack, $tvpo, $tvpd, $tzpo, $tzpd) = @_;
	my ($secs, $usecs) = gettimeofday;

	$tvpd->[$tvpo+0] = $secs;
	$tvpd->[$tvpo+1] = $usecs;
	return 0;
};

#############################################################################
#                                  MATHS                                    #
#############################################################################

$_sin = sub { return sin($_[2]); };
$_cos = sub { return cos($_[2]); };
$_atan = sub { return atan2($_[2], 1); };
$_log = sub { return log($_[2]); };
$_exp = sub { return exp($_[2]); };
$_sqrt = sub { return sqrt($_[2]); };
$_pow = sub { return ($_[2] ** $_[3]); };
